ppcs/ppcTODO-attributes-v2.md

```diff
@@ -0,0 +1,216 @@
+
+```perl
+my $counter :shared;
+
+sub name :lvalue { ... }
+```
+
+The parser supports additional attributes that can be provided by modules, though the situation here has many shortcomings:
+
+* Only lexical or package variables, and subroutines, support attribute syntax. It is not possible to declare attributes on other entities such as packages, or subroutine parameters.
+
+* Attributes on anonymous subroutines are invoked only at "clonecv"-time; the time when an anonymous function gets turned into a closure with variable captures. It cannot perform any behaviour before this time.
+
+* Attribute handling is looked up via the package heirarchy expressed in `@ISA`, rather than by lexical scope.
+
+* Third-party attributes are implemented by providing a single `MODIFY_*_ATTRIBUTES` shouty-named method in a package, which is required to understand all the attributes at once. There is no standard mechanism to create individual attributes independently.
+
+Many of these restrictions come from the way that third-party attributes are provided in the perl core, quite apart from its own built-in handling. Built-in attributes get to run their logic much earlier in the parser.
+
+It is the aim of this specification to provide a better and more flexible way for third-party modules to declare and use attributes to allow authors to declare more interesting behaviours on elements of their code.
+
+## Rationale
+
+(explain why the (following) proposed solution will solve it)
+
+## Specification
+
+Attributes defined by this specification will be lexical in scope, much like that of a `my` variable, `my sub` function, or any of the builtin function exports provided by the `builtin` module. This provides a clean separation of naming within the code.
+
+An attribute is defined primarily by a C callback function, to be invoked by the parser _as soon as_ it has finished parsing the declaration syntax. This callback function will be passed the target (i.e. the item to which the attribute is being attached), and the optional contents of the parentheses used as an argument to the attribute. There is no interesting return value from this callback function.
+
+An optional second callback function can be provided, for the purpose of parsing the incoming text from the source code into a value that the main apply function will use. Experience with the meta-programming layer in `Object::Pad` suggests this is useful, as often when meta-programming it is inconvenience to have to represent the parameters to an attribute as a flat text string.
+
+Pointers to these two functions are found by a structure that is associated with the name of the new attribute, perhaps defined as the following (though exact parameter types are still an open issue; see below):
+
+```c
+struct PerlAttributeDefinition
+{
+    U32 ver;
+    U32 flags;
+    SV * (*parse)(pTHX_ SV *text);   /* optional, may be NULL */
```

Comment:
What does `parse` return? If `parse` isn't supposed to modify the `text` parameter in place, shouldn't it be marked `const`?

Reply:
@mauke good point ... in fact, shouldn't `parse` be rather `Perl_keyword_plugin_t` ?

Reply:
Ahyes, good idea. It could just return the SV itself or it could return some new one. Would have to think a bit about refcounts, but yes it shouldn't modify the one passed in.